## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [31]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [32]:
# View the column names in the data

print(df.columns.values)

['first' 'last' 'job' 'phone' 'email' 'client_id' 'order_id' 'order_date'
 'order_week' 'order_year' 'item_id' 'category' 'subcategory' 'unit_price'
 'unit_cost' 'unit_weight' 'qty' 'line_number']


In [33]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [34]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.describe(include = 'all')
#df.describe(include = 'O')
#df.describe(percentiles = [.05,.1,.9,.95])
#df.corr #not sure how to interpret this and how useful it even is with this data
#df.plot #looks like this unction needs more input to be useful.

first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      order_id  order_date    order_week    order_year  \
count   54639.000000  5.463900e+04       54639  54639.000000  54639.000000   
unique           NaN           NaN         150           NaN           NaN   
top              NaN           NaN  2023-04-19           NaN           NaN   
freq             NaN           NaN         538           NaN           NaN   
mean    54837.869416  5.470190e+06         NaN     11.359139   2022.993064   
std     25487.438231  2.599807e+06         NaN      7.023499      0.082997   
min     10033.000000  1.000886e+06         NaN      1.000000   2022.000000   
25%     33593.000000  3.196372e+06         NaN      6.000000   2023.000000   
50%     53305.000000  5.496966e+06         NaN     11.000000   2023.000000   
75%     78498.000000  7.733869e+06         NaN     17.000000   2023.000000   
max     99984.000000  9.998480e+06         NaN     52.000000   2023.000000   

               item_id     category        subcategory    unit_price  \
count            54639        54639              54639  54639.000000   
unique             500            5                 18           NaN   
top     MAZ00314-80-8M  consumables  bathroom supplies           NaN   
freq               148        23538               6424           NaN   
mean               NaN          NaN                NaN    136.267207   
std                NaN          NaN                NaN    183.873135   
min                NaN          NaN                NaN      0.010000   
25%                NaN          NaN                NaN     20.800000   
50%                NaN          NaN                NaN     68.310000   
75%                NaN          NaN                NaN    173.160000   
max                NaN          NaN                NaN   1396.230000   

           unit_cost   unit_weight           qty   line_number  
count   54639.000000  54639.000000  5.463900e+04  54639.000000  
unique           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN  
mean       99.446073      5.004116  5.702646e+02      2.979667  
std       133.164267      5.326599  1.879552e+04      2.436320  
min         0.010000      0.000000  0.000000e+00      0.000000  
25%        14.840000      1.450000  3.200000e+01      1.000000  
50%        49.890000      3.240000  6.800000e+01      3.000000  
75%       125.570000      6.890000  1.700000e+02      5.000000  
max       846.270000     46.430000  3.958244e+06      9.000000

In [35]:
# What three item categories had the most entries?

df['category'].value_counts().nlargest(3)
#df['category']

consumables    23538
furniture      11915
software        8400
Name: category, dtype: int64

In [36]:
# For the category with the most entries,
# which subcategory had the most entries?
#df.groupby(['category', 'subcategory']).size()

df[df['category'] == 'consumables']['subcategory'].value_counts().nlargest(1)

bathroom supplies    6424
Name: subcategory, dtype: int64

In [37]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().nlargest(5)


33615    220
66037    211
46820    209
38378    207
24741    207
Name: client_id, dtype: int64

In [39]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().nlargest(5).index.tolist()
#print(top_clients)


In [40]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_id = top_clients[0]
df[df['client_id'] == top_client_id]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [41]:
# Create a column that calculates the
# subtotal for each line using the unit_price
# and the qty

#top_2_ids = top_clients[:2]
#print(top_2_ids)

#filter the data to the two clients in question
#top_2_client_data = df[df['client_id'].isin(top_2_ids)]
#top_2_client_data['line_subtotal'] = top_2_client_data['unit_price'] * top_2_client_data['quantity']

# Maybe I misunderstood the directions a bit here. Let's make it simpler and not filter to the two clients already
df['line_subtotal'] = df['unit_price'] * df['qty']

#check if it does what I imagine
df.head()

#ok, so I think the directions were just using the first two rows as an example, not thinking of the two clients specifically.

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [42]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

#df['total_weight'] = df['unit_weight'] * df['qty']
#df['shipping_unit_price'] = df['unit_weight'].apply(lambda x: 7 if x > 50 else 10)
#df['shipping_price'] = df['total_weight'] * df['shipping_unit_price']

# I think the directions here are a little bit wrong; either that, or I don't actually know what the shipping weight is.
# it could also be that the example present within the code is fallacious.
# For now I will just follow the directions as they are on the module assignment paper.

# ok, it looks like the total_weight is what we would actually use to determine the shipping price, rather than the unit weight, as I
# had initially understood.

df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_unit_price'] = df['total_weight'].apply(lambda x: 7 if x > 50 else 10)
df['shipping_price'] = df['total_weight'] * df['shipping_unit_price']

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  total_weight  \
0         7.50  105            1      115164.00        787.50   
1         1.49   21            0         523.95         31.29   
2         1.68   39            6         527.28         65.52   
3         1.23   29            3        1056.18         35.67   
4        46.43   20            1        3902.00        928.60   

   shipping_unit_price  shipping_price  
0                    7         5512.50  
1                   10          312.90  
2                    7          458.64  
3                   10          356.70  
4                    7         6500.20  

[5 rows x 22 columns]

In [43]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['tax'] = (df['line_subtotal'] + df['shipping_price']) * 0.0925
df['line_price'] = df['line_subtotal'] + df['shipping_price'] + df['tax']

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_unit_price  shipping_price  \
0      115164.00        787.50                    7         5512.50   
1         523.95         31.29                   10          312.90   
2         527.28         65.52                    7          458.64   
3        1056.18         35.67                   10          356.70   
4        3902.00        928.60                    7         6500.20   

            tax     line_price  
0  11162.576250  131839.076250  
1     77.408625     914.258625  
2     91.197600    1077.117600  
3    130.691400    1543.571400  
4    962.203500   11364.403500  

[5 rows x 24 columns]

In [47]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   total_weight  shipping_unit_price  shipping_price           tax  \
0        787.50                    7         5512.50  11162.576250   
1         31.29                   10          312.90     77.408625   
2         65.52                    7          458.64     91.197600   
3         35.67                   10          356.70    130.691400   
4        928.60                    7         6500.20    962.203500   

      line_price  line_cost  
0  131839.076250   85597.05  
1     914.258625     629.79  
2    1077.117600     765.18  
3    1543.571400    1077.35  
4   11364.403500    8663.60  

[5 rows x 25 columns]

In [48]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit'] = df['line_price'] - df['line_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...  qty line_number line_subtotal  total_weight  \
0          17        2023  ...  105           1     115164.00        787.50   
1          20        2023  ...   21           0        523.95         31.29   
2           4        2023  ...   39           6        527.28         65.52   
3          17        2023  ...   29           3       1056.18         35.67   
4          21        2023  ...   20           1       3902.00        928.60   

   shipping_unit_price  shipping_price           tax     line_price  \
0                    7         5512.50  11162.576250  131839.076250   
1                   10          312.90     77.408625     914.258625   
2                    7          458.64     91.197600    1077.117600   
3                   10          356.70    130.691400    1543.571400   
4                    7         6500.20    962.203500   11364.403500   

   line_cost   line_profit  
0   85597.05  46242.026250  
1     629.79    284.468625  
2     765.18    311.937600  
3    1077.35    466.221400  
4    8663.60   2700.803500  

[5 rows x 26 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [52]:
# Check your work using the totals above
first_check = df.loc[df['order_id'] == 2742071,'line_price'].sum()
second_check = df.loc[df['order_id'] == 2173913,'line_price'].sum()
third_check = df.loc[df['order_id'] == 6128929,'line_price'].sum()

print(first_check)
print(second_check)
print(third_check)

152811.897075
162388.71929999997
923441.24045


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [63]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_client_spends = [df.loc[df['client_id'] == client, 'line_price'].sum() for client in top_clients]

for i in range(0,5):
  print(top_clients[i],top_client_spends[i])

33615 8377308.572500001
66037 10259514.801275
46820 9743794.320474999
38378 12906550.880874999
24741 82268891.97985001


In [90]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
top_client_df = df[df['client_id'].isin(top_clients)]
aggregated_top_df = top_client_df.groupby('client_id').agg({'qty': 'sum', 'shipping_price': 'sum',
                                                            'line_price': 'sum', 'line_profit': 'sum'})

aggregated_top_df.columns = ["units_purchased","total_shipping_price","total_revenue","total_profit"]
#funny, it looks like client ID does not count as a column as it is used to aggregate
aggregated_top_df


units_purchased  total_shipping_price  total_revenue  total_profit
client_id                                                                    
24741               239862            5126448.37   8.226889e+07  3.657999e+07
33615                64313            1828984.89   8.377309e+06  2.201995e+06
38378                73667            3429455.40   1.290655e+07  3.271830e+06
46820                75768            1601448.84   9.743794e+06  2.736603e+06
66037                43018            1395151.85   1.025951e+07  3.255032e+06

In [110]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

aggregated_top_df.columns = ["Units Purchased","Total Shipping Price","Total Revenue","Total Profit"]
#funny, it looks like client ID does not count as a column as it is used to aggregate
presentation_df = aggregated_top_df.rename_axis("Client ID").reset_index() #this seems to fix the issue

for i in [2,3,4]:
  presentation_df.iloc[:,i]  = (presentation_df.iloc[:,i] / 1000000).apply(lambda x: f'${x:.2f}MM')

presentation_df



Client ID  Units Purchased Total Shipping Price Total Revenue Total Profit
0      24741           239862              $5.13MM      $82.27MM     $36.58MM
1      33615            64313              $1.83MM       $8.38MM      $2.20MM
2      38378            73667              $3.43MM      $12.91MM      $3.27MM
3      46820            75768              $1.60MM       $9.74MM      $2.74MM
4      66037            43018              $1.40MM      $10.26MM      $3.26MM

In [111]:
# Sort the updated data by "Total Profit" form highest to lowest
df_sorted = presentation_df.sort_values('Total Profit', ascending=False)
df_sorted

Client ID  Units Purchased Total Shipping Price Total Revenue Total Profit
0      24741           239862              $5.13MM      $82.27MM     $36.58MM
2      38378            73667              $3.43MM      $12.91MM      $3.27MM
4      66037            43018              $1.40MM      $10.26MM      $3.26MM
3      46820            75768              $1.60MM       $9.74MM      $2.74MM
1      33615            64313              $1.83MM       $8.38MM      $2.20MM